akka-distributed-data/src/main/scala/akka/cluster/ddata/GCounter.scala

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.cluster.ddata
+
+import akka.cluster.Cluster
+import akka.cluster.UniqueAddress
+
+object GCounter {
+  val empty: GCounter = new GCounter
+  def apply(): GCounter = empty
+  /**
+   * Java API
+   */
+  def create(): GCounter = empty
+
+  /**
+   * Extract the [[GCounter#value]].
+   */
+  def unapply(c: GCounter): Option[Long] = Some(c.value)
+}
+
+/**
+ * Implements a 'Growing Counter' CRDT, also called a 'G-Counter'.
+ *
+ * It is described in the paper
+ * <a href="http://hal.upmc.fr/file/index/docid/555588/filename/techreport.pdf">A comprehensive study of Convergent and Commutative Replicated Data Types</a>.
+ *
+ * A G-Counter is a increment-only counter (inspired by vector clocks) in
+ * which only increment and merge are possible. Incrementing the counter
+ * adds 1 to the count for the current node. Divergent histories are
+ * resolved by taking the maximum count for each node (like a vector
+ * clock merge). The value of the counter is the sum of all node counts.
+ *
+ * This class is immutable, i.e. "modifying" methods return a new instance.
+ */
+@SerialVersionUID(1L)
+final class GCounter private[akka] (
+  private[akka] val state: Map[UniqueAddress, Long] = Map.empty)
```

Comment:
BigInteger is a plain Java class, BigInt just a Scala wrapper. Serialization would of course cost proportional to the actual size, which most of the time will be small. BigInteger provides serialization to `byte[]` itself.